appengine/flexible/TwelveFactor/Startup.cs

```diff
@@ -0,0 +1,172 @@
+        {
+            var projectId = GetProjectId();
+
+            // Only use Console and Debug logging during development.
+            loggerFactory.AddConsole(Configuration.GetSection("Logging"));
+            loggerFactory.AddDebug();
+            // Send logs to Stackdriver Logging.
+            loggerFactory.AddGoogle(projectId);
+            
+            _cloudStorage.Logger = loggerFactory.CreateLogger<CloudStorageFileProvider>();
+
+            if (env.IsDevelopment())
+            {
+                app.UseDeveloperExceptionPage();
+                app.UseBrowserLink();
+            }
+            else
+            {
+                app.UseExceptionHandler("/Home/Error");
+                app.UseGoogleExceptionLogging();
+            }
+
+            app.UseStaticFiles();
+            app.UseGoogleTrace();
+
+            app.UseMvc(routes =>
+            {
+                routes.MapRoute(
+                    name: "default",
+                    template: "{controller=Home}/{action=Index}/{id?}");
+            });
+        }
+
+        private string GetProjectId()
+        {
+            var instance = Google.Api.Gax.Platform.Instance();
+            var projectId = instance?.ProjectId ?? Configuration["Google:ProjectId"];
+            if (string.IsNullOrEmpty(projectId))
+            {
+                throw new Exception(
+                    "The logging, tracing and error reporting libraries need a project ID. " +
+                    "Update appsettings.json by setting the ProjectId property with your " +
+                    "Google Cloud Project ID, then recompile.");
+            }
+            return projectId;
+        }
+
+        private string GetServiceName()
+        {
+            var instance = Google.Api.Gax.Platform.Instance();
+            // An identifier of the service. See https://cloud.google.com/error-reporting/docs/formatting-error-messages#FIELDS.service.
+            var serviceName =
+                instance?.GaeDetails?.ServiceId ??
+                Configuration["Google:ErrorReporting:ServiceName"];
+            if (string.IsNullOrEmpty(serviceName))
+            {
+                throw new Exception(
+                    "The error reporting library needs a service name. " +
+                    "Update appsettings.json by setting the Google:ErrorReporting:ServiceName property with your " +
+                    "Service Id, then recompile.");
+            }
+            return serviceName;
+        }
+
+        private string GetVersion()
+        {
+            var instance = Google.Api.Gax.Platform.Instance();
+            // The source version of the service. See https://cloud.google.com/error-reporting/docs/formatting-error-messages#FIELDS.version.
+            var versionId =
+                instance?.GaeDetails?.VersionId ??
+                Configuration["Google:ErrorReporting:Version"];
+            if (string.IsNullOrEmpty(versionId))

```

Comment:
This is quite harsh isn't it? Why do we need to crash the app if there's no version? We should have a suitable default that can be found in the logs/traces easily. Show a warning that the version is not set.